### ISSUE TITLE
Implement stellar uncertainties

### ISSUE THREAD

**Issue by tcrundall** [tcrundall/chronostar-tng#24]

**Comment by tcrundall** [comment]
Two potential approaches.

1. Let users provide covariance matrices as extra input.
cons: this adds complexity to the interface. I'm trying to keep front end as simple as possible. This also adds ambiguity to how I interface with scikit-learn's mixture models.

2. Have covariance matrices encoded into the data array.
cons: Have to reconstruct covariance matrices every iteration. (though, since this is immediately followed by complex overlap integration, this is perhaps not too big a deal)

I elect for the second approach.

**Comment by tcrundall** [comment]
Another point of consideration is, how to do component parameter maximisation with stellar uncertainties.

I'd like to retain the performance improvement achieved by only exploring the age, and then deriving the mean and covariance matrix from that.

### Cov-free approach
One approach could be: sample the star covariances ~100 times each (perhaps scaled dynamically by dimension of space), and then perform same routine which disregards covariance matrices.

I like this because it lends itself to reuse of code.

But one must consider correctness and performance:
- is doing the cov-free approach 100x faster than doing an overlap integral?
- is 100 random samples sufficiently accurate? (introduces some stochasticity too which is perhaps undesirable)

### OG Chronostar
An alternative approach is to do a full parameter exploration of the component's parameter space: mean (6 pars), age (1 par) and covariance matrix (2 - 21 pars). One would also have to specifically implement which parameters are being explored and how those parameters build the covariance matrix. Truly a nightmare.

### Hybrid
A hybrid approach could be to use the cov-free approach to get close to the solution, and then do a complicated full-parameter exploration to fine-tune the parameters.

### Decision
For simplicity, I am electing to go for the cov-free approach.

**Comment by mikeireland** [comment]
Tim - the "Cov-free" approach really is not physical and should not work. e.g. just because stars in a cluster have uncertain distances, it doesn't mean the cluster is elongated in the distance direction. Covariance matrices will _always_ be elongated in at least one direction (the radial velocity). Incorporating this into Chronostar was always anticipated to be a strength.

**Comment by mikeireland** [comment]
Some numbers: for a typical distance of 200pc, at G=13, we anticipate 5 micro-arcsec/year proper motions at Gaia DR5, which corresponds to 5 m/s. Compared to 1km/s for the radial velocity, and ~500m/s for the intrinsic velocity dispersion of a component. The 7 micro-arcsec parallax error corresponds to 0.3pc, which is admittedly not significant compared to the few pc minimum initial group size. But for fainter or more distant components, distance errors are again significant. Radial velocities will not exist for a great many stars we want to base a cluster model on.

**Comment by tcrundall** [comment]
Noted.

I acknowledge that fitting this way will "bloat" the component along axes of large uncertainty. I had overlooked that so thanks for pointing that out.

Though it's not obvious to me that this will result in super unrealistic components since any artificial bloating might be countered by intrinsic phase-space constraints. i.e. a component's velocity spread in the RV direction must remain consistent with the current day position spread. I note that this relationship is degenerate with the age, which is bad, but the two other phase-space planes would serve to constrain the age.

I think it's worth pursuing the simpler approach first - it aligns with my overall approach of producing a simple, light-weight chronostar into which more realistic, heavy-weight modules can be substituted in later. And who knows, it might actually yield surprisingly useful results. At the very least it should do a decent enough job at establishing component members, and then these components could be fit to in isolation to get a definite age.

**Comment by tcrundall** [comment]
Upon further reflection, I'm not sure what the difference would be.

The "Cov-free" approach is effectively doing a discrete (albeit stochastic) overlap integral.

The issue you pointed out, about a component not being elongated along some axis just because all its members are uncertain, is present in OG (paper 1 and 2) Chronostar too.

So the only differences OG Chronostar has are: 1) more accurate overlap integral, and 2) exploring all component parameters compared to just age.

**Comment by mikeireland** [comment]
I am near certain that "cov-free" is not doing a discrete overlap integral. So sampling the stellar covariance matrices is a tangent that I don't believe is publishable. So please prioritise everything else. We can meet with Mark about the math later. I do agree that the point samples have some use, but working with data that have no RV measurements or highly uncertain ones is much more important.

**Comment by tcrundall** [comment]
Acknowledged. I'll consider it an unpublishable stepping stone. FYI the cov-free is already implemented (it took less than an hour) so I won't be putting any more time into it.

I think an in depth discussion about missing/uncertain data is needed. Talking with Javier and Marusa,

**Comment by tcrundall** [comment]
Sorry, I couldn't help myself. We can say I did this off the clock if you like.

This is comparing relative error of computing an overlap integral by random sampling. The commit above has the script I used.

![image](https://user-images.githubusercontent.com/8383244/194775993-723e15d7-6941-405a-bdce-c014a98dd73e.png)

legend is dimension
`nstars = 1_000_000`
The average relative error is ~14% for 100 samples, and 1% for 10_000 samples.
